dynawo/sources/Models/CPP/Controls/Shunts/ModelCentralizedShuntsSectionControl/DYNModelCentralizedShuntsSectionControl.cpp

```diff
@@ -0,0 +1,369 @@
+  ModelCentralizedShuntsSectionControl::defineVariables(std::vector<boost::shared_ptr<Variable> >& variables) {
+    // only one UMonitoredPu variable shared among all shunts as all of them are regulating the same node
+    variables.push_back(VariableNativeFactory::createState("UMonitoredPu_value", CONTINUOUS));
+    std::stringstream URefPuName;
+    for (int s = 0; s < nbShunts_; ++s) {
+      URefPuName.str(std::string());
+      URefPuName.clear();
+      URefPuName << "URefPu_" << s << "_value";
+      variables.push_back(VariableNativeFactory::createState(URefPuName.str(), DISCRETE));
+    }
+    std::stringstream sectionName;
+    for (int s = 0; s < nbShunts_; ++s) {
+      sectionName.str(std::string());
+      sectionName.clear();
+      sectionName << "section_" << s << "_value";
+      variables.push_back(VariableNativeFactory::createState(sectionName.str(), DISCRETE));
+    }
+  }
+
+  void
+  ModelCentralizedShuntsSectionControl::init(const double /*t0*/) {
+    /* not need */
+  }
+
+  void
+  ModelCentralizedShuntsSectionControl::collectSilentZ(BitMask* /*silentZTable*/) {
+    /* not need */
+  }
+
+  void
+  ModelCentralizedShuntsSectionControl::getSize() {
+    sizeF_ = 0;
+    sizeY_ = 1;  // UMonitoredPu
+    sizeZ_ = 2 * nbShunts_;  // URefPu values are stored at indexes [0, nbShunts[, sections are stored at [nbShunts, 2 * nbShunts[
+    sizeG_ = 4;
+    sizeMode_ = 0;
+    calculatedVars_.assign(nbCalculatedVariables_, 0);
+  }
+
+  void
+  ModelCentralizedShuntsSectionControl::evalStaticYType() {
+    yType_[0] = EXTERNAL;
+  }
+
+  void
+  ModelCentralizedShuntsSectionControl::evalF(double /*t*/, propertyF_t /*type*/) {
+    /* not need */
+  }
+
+  void
+  ModelCentralizedShuntsSectionControl::evalJt(const double /*t*/, const double /*cj*/, SparseMatrix& /*jt*/, const int /*rowOffset*/) {
+    /* not need */
+  }
+
+  void
+  ModelCentralizedShuntsSectionControl::evalG(const double t) {
+    double UMonitoredPu = yLocal_[0];
+    double minValue;
+    double maxValue;
+    if (!isSelf_) {
+      for (int s = 0; s < nbShunts_; ++s) {
+        minValue = zLocal_[s] - DeadbandsUPu_[s];
+        maxValue = zLocal_[s] + DeadbandsUPu_[s];
+        if (doubleNotEquals(UMonitoredPu, minValue) &&
+          UMonitoredPu < minValue &&
+          sections0_[s] < SectionsMax_[s]) {
+          gLocal_[0] = ROOT_UP;
+          gLocal_[1] = ROOT_DOWN;
+          gLocal_[2] = ((t - whenUp_) > tNext_ || doubleEquals((t - whenUp_), tNext_)) ? ROOT_UP : ROOT_DOWN;
+          gLocal_[3] = ROOT_DOWN;
+          changingShunt = s;
+          break;
+        } else if (doubleNotEquals(UMonitoredPu, maxValue) &&
+          UMonitoredPu > maxValue &&
+          sections0_[s] > SectionsMin_[s]) {
+          gLocal_[0] = ROOT_DOWN;
+          gLocal_[1] = ROOT_UP;
+          gLocal_[2] = ROOT_DOWN;
+          gLocal_[3] = ((t - whenDown_) > tNext_ || doubleEquals((t - whenDown_), tNext_)) ? ROOT_UP : ROOT_DOWN;
+          changingShunt = s;
```

Comment:
I have modified z so that we store Uref at index 0: mix and max values are then always centered around same value. A break statement has been added too